ejb3/src/main/java/org/jboss/as/ejb3/tx/CMTTxInterceptor.java

```diff
@@ -252,41 +175,68 @@ protected Object invokeInCallerTx(InterceptorContext invocation, Transaction tx,
         try {
             return invocation.proceed();
         } catch (Throwable t) {
-            handleInCallerTx(invocation, t, tx, component);
+            ApplicationExceptionDetails ae = component.getApplicationException(t.getClass(), invocation.getMethod());
+
+            if (ae != null) {
+                if (ae.isRollback()) setRollbackOnly(tx);
+                // an app exception can never be an Error
+                throw (Exception) t;
+            }
+            try {
+                throw t;
+            } catch (EJBTransactionRolledbackException | NoSuchEJBException | NoSuchEntityException e) {
+                setRollbackOnly(tx);
+                throw e;
+            } catch (RuntimeException e) {
+                setRollbackOnly(tx);
+                throw new EJBTransactionRolledbackException(e.getMessage(), e);
+            } catch (Error e) {
+                setRollbackOnly(tx);
+                throw EjbLogger.ROOT_LOGGER.unexpectedErrorRolledBack(e);
+            }
         }
-        throw new RuntimeException("UNREACHABLE");
     }
 
     protected Object invokeInNoTx(InterceptorContext invocation, final EJBComponent component) throws Exception {
         try {
             return invocation.proceed();
-        } catch (Throwable t) {
-            handleExceptionInNoTx(invocation, t, component);
+        } catch (EJBException | RemoteException e) {
+            throw e;
+        } catch (RuntimeException e) {
+            throw new EJBException(e);
+        } catch (Exception e) {
+            throw e;
+        } catch (Error e) {
+            ApplicationExceptionDetails ae = component.getApplicationException(e.getClass(), invocation.getMethod());
```

Comment:
There is a comment a few lines above that says '// an app exception can never be an Error', so either this or the comment would appear to be incorrect.